cmd/all-in-one/main.go

```diff
@@ -130,14 +131,25 @@ by default uses only in-memory database.`,
 			})
 			c.Start(cOpts)
 
-			startAgent(aOpts, repOpts, tchanBuilder, grpcBuilder, cOpts.CollectorGRPCPort, logger, metricsFactory)
+			// agent
+			grpcBuilder.CollectorHostPorts = append(grpcBuilder.CollectorHostPorts, fmt.Sprintf("127.0.0.1:%d", cOpts.CollectorGRPCPort))
+			agentMetricsFactory := metricsFactory.Namespace(metrics.NSOptions{Name: "agent", Tags: nil})
+			cp, err := agentApp.CreateCollectorProxy(repOpts, tchanBuilder, grpcBuilder, logger, agentMetricsFactory)
+			if err != nil {
+				logger.Fatal("Could not create collector proxy", zap.Error(err))
+			}
+			agent := startAgent(cp, aOpts, logger, metricsFactory)
```

Comment:
If we had a notion of life cycle (e.g. https://godoc.org/go.uber.org/fx#Lifecycle), this code could've been moved to a shared component, because all the individual objects created here are _only_ exposed to be able to call Close() on some of them.  Our internal builds using Fx extensively, making `main()` 10-lines long, while having all initialization code fully testable (unlike main). But DI in Go is a tricky proposition (e.g. this pearl: https://github.com/uber-go/fx/issues/694).

We don't have to do full DI though, the `svc.RunAndThen` below is already a form of lifecycle management. We could extend it with just OnStart/OnStop hooks.

Reply:
DI is indeed one of the things I miss from Java. Let's see how the code here looks in the end, and discuss adopting a better pattern for all components. This might bring benefits if we decide to make the components more easily consumable as a package (like, from [jaeger-opentelemetry-collector](https://github.com/jaegertracing/jaeger-opentelemetry-collector))